### ISSUE TITLE
hdf5 can not be pickled with neither store nor to_hdf5 method

### ISSUE THREAD

**Issue by Runkli** [dask/dask#6482]
The code which generated this issue is:
```python
procDask = da.stack(procArrays)
print(procDask)
print(procDask.compute())    
    
#da.to_hdf5('out.hdf5',{'/arr',procDask})
    
f = h5py.File('out.hdf5',mode='w')
d = f.require_dataset('/arr', shape=procDask.shape, dtype=procDask.dtype)
da.store(procDask, d)
```
Where procArrays are some 3x3 numpy arrays
My ```procDask``` shows to be:
```
dask.array<stack, shape=(3, 3, 3), dtype=int16, chunksize=(1, 3, 3), chunktype=numpy.ndarray>
```
However using any of the above-mentioned writing methods (both the commented and uncommented one), gives me the following error:
```
TypeError: h5py objects cannot be pickled
```
**Environment**:

- Dask version: 2.11.0
- Python version: 3.7.6
- Operating System: 20.04
- Install method (conda, pip, source): conda

**Comment by gforsyth** [comment]
Hi, thanks for reporting!  Can you please put together a minimum reproducible example so that the maintainers can more easily find the bug?

You can read https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports for more info.

**Comment by Runkli** [comment]
I got the error: ```TypeError: unhashable type: 'Array'``` from the following code.
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-from dask.distributed import Client
import dask.array as da
from dask import delayed
from dask.distributed import Client
import numpy as np

def main():
    procArray = da.ones((3,3,3))
    da.to_hdf5('out.hdf5',{'/arr',procArray})
    client.shutdown()
    
if __name__ == "__main__":
	client = Client()
	main()
```
Instead if I try to store the ```procArray.compute()``` I get: ```TypeError: unhashable type: 'numpy.ndarray'```

**Comment by m-albert** [comment]
Hey @Runkli ,

in the code you posted, the error comes from a wrong call to `to_hdf5`. See this modification:
```python
def main():
    procArray = da.ones((3,3,3))
#     da.to_hdf5('out.hdf5',{'/arr',procArray})
    procArray.to_hdf5('out.hdf5','/arr')
    client.shutdown()
```

However, the code would still fail and this is related to the issue I just opened https://github.com/dask/dask/issues/6522.

Your code runs through if you don't create a `dask.distributed` client:

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-from dask.distributed import Client
import dask.array as da
#from dask import delayed
#from dask.distributed import Client
import numpy as np

def main():
    procArray = da.ones((3,3,3))
#     da.to_hdf5('out.hdf5',{'/arr',procArray})
    procArray.to_hdf5('out.hdf5','/arr')
    client.shutdown()
    
if __name__ == "__main__":
	client = Client()
	main()
```

**Comment by Runkli** [comment]
Exactly! I forgot to update the issue here, but I am stuck exactly at your issue also. I can make the call just fine without the distributed client, but when I initialize it, I get this error. As I understand hdf5 files can not be serialized, which can not be used in a distributed form, but wouldn't getting rid of the client make it essentially serial? Or is that only needed for large clusters?

**Comment by m-albert** [comment]
By default, for `dask.array` a scheduler is used that parallelizes tasks over threads (according to [here](https://docs.dask.org/en/latest/scheduler-overview.html#configuring-schedulers)). So even without using `dask.distributed` your code could (depends on whether your tasks release the GIL) run faster than if using the (serial) 'single-threaded' scheduler.

**Comment by GenevieveBuckley** [comment]
Closing this issue, at it's a duplicate of https://github.com/dask/dask/issues/2488 (which Marvin's issue https://github.com/dask/dask/issues/6522 is also a duplicate of). Progress can be tracked over there.
